src/devices/ec3k.c

```diff
@@ -0,0 +1,317 @@
+    const double energy2_kWh = energy2 / (1000.0 * 3600.0);   // Ws to kWh
+    uint32_t time_total      = (uint32_t)time_total_low | ((uint32_t)time_total_high << 16);
+    uint32_t time_on         = (uint32_t)time_on_low | ((uint32_t)time_on_high << 16);
+
+    if (pad_1 == 0 && pad_2 == 0 && pad_3 == 0 && pad_4 == 0) {
+        if (calculated_crc == received_crc) {
+            /* clang-format off */
+            data_t *data = data_make(
+                "model",            "",              DATA_STRING, "Voltcraft-EnergyCount3000",
+                "id",               "",              DATA_FORMAT, "%04x", DATA_INT, id,
+                "power",            "Power",         DATA_DOUBLE, power_current,
+                "energy",           "Energy",        DATA_DOUBLE, energy_kWh,
+                "energy2",          "Energy 2",      DATA_DOUBLE, energy2_kWh,
+                "mic",              "Integrity",     DATA_STRING, "CRC",
+                "time_total",       "Time total",    DATA_INT,    time_total,
+                "time_on",          "Time on",       DATA_INT,    time_on,
+                "power_max",        "Power max",     DATA_DOUBLE, power_max,
+                "reset_counter",    "Reset counter", DATA_INT,    reset_counter,
+                "flags",            "Flags",         DATA_INT,    flags,
+                NULL);
+            /* clang-format on */
+
+            decoder_output_data(decoder, data);
+            rc = 1;
+        }
+        else {
+            decoder_logf(decoder, 1, __func__, "Warning: CRC error, calculated %04X but received %04X", calculated_crc, received_crc);
+            rc = DECODE_FAIL_MIC;
+        }
+    }
+    else {
+        decoder_logf(decoder, 1, __func__, "Warning: padding bits are not zero, pad_1=%u pad_2=%u pad_3=%u pad_4=%u", pad_1, pad_2, pad_3, pad_4);
+        rc = DECODE_FAIL_SANITY;
+    }
+
+    return rc;
+}
+
+// copied from the ec3k python implementation at https://github.com/avian2/ec3k
+static inline uint16_t calc_ec3k_crc(const uint8_t *buffer, size_t len)
+{
+    uint16_t crc = 0xffff;
+    for (size_t i = 0; i < len; i++) {
+        uint8_t ch = buffer[i];
+        ch ^= crc & 0xff;
+        ch ^= (ch << 4) & 0xff;
+        crc = (((uint16_t)ch << 8) | (crc >> 8)) ^ ((uint16_t)ch >> 4) ^ ((uint16_t)ch << 3);
+    }
+    return crc;
+}
+
+/*
+ * List of fields that may appear in the output
+ *
+ * Used to determine what fields will be output in what
+ * order for this device when using -F csv.
+ *
+ */
+static char const *const output_fields[] = {
+        "model",
+        "id",
+        "power",
+        "energy",
+        "energy2",
+        "time_total",
+        "time_on",
+        "power_max",
+        "reset_counter",
+        "flags",
+        "mic",
+        NULL,
+};
+
+r_device const ec3k = {
+        .name        = "Voltcraft-EnergyCount3000",
+        .modulation  = FSK_PULSE_PCM,
+        .short_width = BITTIME_US,
```

Comment:
Use 50 here and remove the BITTIME_US constant.

Reply:
> Use 50 here and remove the BITTIME_US constant.

that really hurts me ;-) We tell our developers never to use magic constants. Hopefully none of my developers ever sees that I have done that.

I understand that source code now more like a macro language for symbolizer.py which can also be feed into a C compiler.

I need to check your feedback later, I don't know if I will have time the next days.